### PR TITLE
feat: add eth_getTransactionByBlockHashAndIndex and eth_getTransactionByBlockNumberAndIndex @nbaztec

### LINKED ISSUE #46: eth_api: Implement `get_transaction_by_block_hash_and_index` method
### **Description:**
We need to implement the [`get_transaction_by_block_hash_and_index`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L1051C8-L1051C47) method. This function is designed to retrieve a specific transaction from a block, given the block's hash and the transaction's index within that block.

```rust
/// Fetches a transaction from a block based on its hash and its position index.
fn get_transaction_by_block_hash_and_index(&self, block_hash: H256, index: U256) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<Option<Transaction>>> {
    not_implemented("get_transaction_by_block_hash_and_index")
}
```

#### Expected response
Command:
```
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_getTransactionByBlockHashAndIndex",
       "params": ["0xfb8718465e855a68e2603b0e21c1b974288d4628419792aa89b68e461666934b", "0x0"]
     }'
```
Response:
```
{"jsonrpc":"2.0","result":,"id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - This function will accept a `H256` block hash and a `U256` index representing the transaction's position within the block.
    - It should return an `Option<Transaction>` which represents the specific transaction, or `None` if it doesn't exist.

2. **Unit Tests:**
    - Write unit tests to ensure the `get_transaction_by_block_hash_and_index` method is working accurately.
    - Take into account various scenarios and edge cases, including non-existent transaction indices.
    - Ensure that all possible errors, such as an invalid block hash or index, are properly addressed.

3. **Inline Documentation:**
    - Add inline Rust documentation (`///` comments) explaining the function and its usage.
    - Include any relevant information about the return values and potential errors/exceptions.

### **Task Checklist:**
- [ ] Implement the `get_transaction_by_block_hash_and_index` method.
- [ ] Write unit tests for the `get_transaction_by_block_hash_and_index` method.
- [ ] Add inline documentation for the `get_transaction_by_block_hash_and_index` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### PR DESCRIPTION
**Pull request by nbaztec**
# What :computer: 
* Adds `eth_getTransactionByBlockHashAndIndex` and `eth_getTransactionByBlockNumberAndIndex`
* `testing::apply_tx()` now also returns the block number the tx was included in, along with the block's hash.

# Why :hand:
* Methods are used to fetch transactions as basic functionality.

# Evidence :camera:
image

# Notes :memo:
* Fixes #46 , #47

### CODE REVIEW

File: SUPPORTED_APIS.md [hunk 1]
```diff
@@ -48,8 +48,8 @@ The `status` options are:
 | [`ETH`](#eth-namespace) | [`eth_getLogs`](#eth_getlogs) | `SUPPORTED` | Returns an array of all logs matching a given filter object |
 | `ETH` | `eth_getProof` | `NOT IMPLEMENTED` | Returns the details for the account at the specified address and block number, the account's Merkle proof, and the storage values for the specified storage keys with their Merkle-proofs |
 | [`ETH`](#eth-namespace) | `eth_getStorageAt`(#`eth_getstorageat) | `SUPPORTED` | Returns the value from a storage position at a given address |
```
Comment (MexicanAce) [inline SUPPORTED_APIS.md:50]: Oh, just noticed this entry was malformed from the last PR
```suggestion
| [`ETH`](#eth-namespace) | [`eth_getStorageAt`](#eth_getstorageat) | `SUPPORTED` | Returns the value from a storage position at a given address |
```

File: SUPPORTED_APIS.md [hunk 2]
```diff
@@ -1101,6 +1100,65 @@ curl --request POST \
...
+
+```bash
+curl --request POST \
+  --url http://localhost:8011/ \
+  --header 'content-type: application/json' \
+  --data '{
+    "jsonrpc": "2.0",
+    "id": "1",
+    "method": "eth_getTransactionByBlockNumberAndIndex",
+    "params": ["atest", "0x1"]
```
Comment (MexicanAce) [inline SUPPORTED_APIS.md:1157]: typo: 
```suggestion
    "params": ["latest", "0x1"]
```

File: src/node.rs [hunk 3]
```diff
@@ -4232,4 +4349,319 @@ mod tests {
...
+            .get_transaction_by_block_number_and_index(
+                BlockNumber::Number(input_block_number),
+                U64::from(1),
+            )
+            .await
+            .expect("failed fetching transaction")
+            .expect("no transaction");
+
+        assert_eq!(input_tx_hash, actual_tx.hash);
+        assert_eq!(Some(U64::from(100)), actual_tx.block_number);
```
Comment (MexicanAce) [inline src/node.rs:4665]: nit/suggestion: Could be updated to use the variable from earlier
```suggestion
        assert_eq!(Some(input_block_number), actual_tx.block_number);
```

**Review by MexicanAce — APPROVED** [review]
LGTM
